### PR TITLE
Alerting: Send configuration and state to the remote Alertmanager on shutdown

### PR DESCRIPTION
**Pull request by santihernandezc**
When Grafana shuts down, `StopAndWait()` is called on the Alertmanager struct.
With this PR if Grafana is running in _remote secondary_ mode, the forked Alertmanager calls `ApplyConfig()` on the remote Alertmanager struct as part of the shutdown process to send the latest state and configuration to the remote Alertmanager.

### CODE REVIEW

File: pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go [hunk 1]
```diff
@@ -93,8 +137,26 @@ func (fam *RemoteSecondaryForkedAlertmanager) CleanUp() {
...
+	}
+
+	config, err := fam.store.GetLatestAlertmanagerConfiguration(ctx, fam.orgID)
+	if err != nil {
+		fam.log.Error("Error getting latest Alertmanager configuration", "err", err)
+		return
+	}
+	if err := fam.remote.CompareAndSendConfiguration(ctx, config); err != nil {
+		fam.log.Error("Error sending configuration to the remote Alertmanager", "err", err)
+	}
```
Comment (gotjosh) [inline pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:159]: I don't think we need this to be a comparison - shouldn't we just send whatever we have at this point? What's your reasoning for the comparison?

Comment (santihernandezc) [reply pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:193]: 1. If we have the same on both Alertmanagers, there's no need to create a new configuration in the remote Alertmanager. The same goes for the state.
2. We'd need to add public methods for sending the configuration without checking first.

Ultimately I think this should follow the same logic as `ApplyConfig()`, I don't see any difference between what we want to achieve here and on every call to `ApplyConfig()`. If we were to remove the comparison here, we should do the same in the other method.

Comment (santihernandezc) [reply pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:193]: I created a [PR to remove this comparison](https://github.com/grafana/grafana/pull/79469), I don't think it makes much sense to keep it around. We can use that change in this PR once it's merged.

File: pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go [hunk 2]
```diff
@@ -160,9 +171,26 @@ func (fam *RemoteSecondaryForkedAlertmanager) CleanUp() {
...
 
 func (fam *RemoteSecondaryForkedAlertmanager) StopAndWait() {
+	// Stop the internal Alertmanager.
 	fam.internal.StopAndWait()
+	// Stop our alert senders.
 	fam.remote.StopAndWait()
-	// TODO: send config and state on shutdown.
+
+	// Send config and state to the remote Alertmanager.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (gotjosh) [inline pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:180]: I don't think we should timeout here -- basically we want to allow a Grafana pod as much as it needs to ensure the state is saved (we have a global timeout of 30s anyways)

File: pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go [hunk 3]
```diff
@@ -160,9 +171,26 @@ func (fam *RemoteSecondaryForkedAlertmanager) CleanUp() {
...
+	defer cancel()
+	if err := fam.remote.CompareAndSendState(ctx); err != nil {
+		fam.log.Error("Error sending state to the remote Alertmanager", "err", err)
```
Comment (gotjosh) [inline pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:183]:
```suggestion
		fam.log.Error("Error sending state to the remote Alertmanager while stopping", "err", err)
```

File: pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go [hunk 4]
```diff
@@ -160,9 +171,26 @@ func (fam *RemoteSecondaryForkedAlertmanager) CleanUp() {
...
+	}
+
+	config, err := fam.store.GetLatestAlertmanagerConfiguration(ctx, fam.orgID)
+	if err != nil {
+		fam.log.Error("Error getting latest Alertmanager configuration", "err", err)
```
Comment (gotjosh) [inline pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:188]:
```suggestion
		fam.log.Error("Error getting latest Alertmanager configuration while stopping", "err", err)
```

File: pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go [hunk 5]
```diff
@@ -160,9 +171,26 @@ func (fam *RemoteSecondaryForkedAlertmanager) CleanUp() {
...
+		return
+	}
+	if err := fam.remote.CompareAndSendConfiguration(ctx, config); err != nil {
+		fam.log.Error("Error sending configuration to the remote Alertmanager", "err", err)
```
Comment (gotjosh) [inline pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go:192]:
```suggestion
		fam.log.Error("Error sending configuration to the remote Alertmanager while stopping", "err", err)
```

**Review by gotjosh — APPROVED** [review]
LGTM but please see my comments.

**Pull request merged**